File: track-b-data-engineering/dagster_project/ai/provider.py
"""ILLMProvider interface — mirrors track-a-jd-native/src/ai/provider.ts."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Protocol

EnrichmentField = Literal[
    "translate_cn_to_en",
    "extract_callouts",
    "infer_make",
]


@dataclass(frozen=True)
class EnrichmentRequest:
    id: str
    field: EnrichmentField
    inputs: dict[str, str | None]


@dataclass(frozen=True)
class EnrichmentMeta:
    provider: str
    prompt_template_ver: str
    tokens_in: int | None = None
    tokens_out: int | None = None
    cost_usd: float | None = None
    latency_ms: int | None = None
    cache_hit: bool = False


@dataclass(frozen=True)
class EnrichmentResponse:
    id: str
    field: EnrichmentField
    result: str | list[int] | None
    meta: EnrichmentMeta
    confidence: Literal["high", "medium", "low"] | None = None


class ILLMProvider(Protocol):
    """Single interface every provider implements."""

    @property
    def name(self) -> str: ...

    async def enrich(self, req: EnrichmentRequest) -> EnrichmentResponse: ...

File: track-a-jd-native/src/ai/provider.ts
/**
 * LLM provider abstraction.
 *
 * Every LLM call goes through this interface. Implementations:
 *   • MockProvider — fixture-backed, used in tests + as the universal
 *                    fallback so the reviewer never sees a missing-cache
 *                    crash.
 *   • CachedProvider — decorator wrapping any upstream; SQLite-backed.
 *                      Default for reviewer runs (LLM_PROVIDER=cached).
 *   • ClaudeCodeHandoffProvider — emits task files for the operator
 *                                 (me) to translate via the Claude Max
 *                                 session, then reads results back.
 *                                 Used to seed the cache.
 *   • OllamaProvider, GeminiFreeProvider, AnthropicBatchProvider —
 *                      production-target implementations (stubs in this
 *                      submission). See ADR-007.
 */

/** A field needing enrichment. */
export type EnrichmentField = "translate_cn_to_en" | "extract_callouts" | "infer_make";

export interface EnrichmentRequest {
  /** Stable opaque task id used by the handoff provider. */
  id: string;
  field: EnrichmentField;
  inputs: {
    /** For translate_cn_to_en: the Chinese text. */
    cn?: string;
    /** For extract_callouts: R2 key of the schematic image. */
    image_r2_key?: string;
    /** For infer_make: model code. */
    model_code?: string;
  };
  /** Optional context the model may use for better answers. */
  context?: Record<string, unknown>;
}

export interface EnrichmentResponse {
  id: string;
  field: EnrichmentField;
  /** Result is the natural output for the field — string for translate/infer, number[] for callouts. */
  result: string | number[] | null;
  /** Confidence, if the provider can supply it. */
  confidence?: "high" | "medium" | "low";
  /** Provider metadata for audit. */
  meta: {
    provider: string;
    promptTemplateVer: string;
    tokensIn?: number;
    tokensOut?: number;
    costUsd?: number;
    latencyMs?: number;
    cacheHit?: boolean;
  };
}

export interface ILLMProvider {
  readonly name: string;
  enrich(req: EnrichmentRequest): Promise<EnrichmentResponse>;
}
